torch/csrc/autograd/python_cpp_function.cpp

```diff
@@ -173,22 +173,27 @@ PyObject* functionToPyObject(std::shared_ptr<Function> cdata)
     return obj;
   }
 
-  auto& fn = *cdata;
-  auto it = cpp_function_types.find(std::type_index(typeid(fn)));
-  if (it == cpp_function_types.end()) {
-    return PyErr_Format(PyExc_TypeError,
-        "Don't know how to create Python object for %s", typeid(fn).name());
-  }
+  if (cdata->pyobj) {
+    Py_INCREF(cdata->pyobj);
+  } else {
+    auto& fn = *cdata;
+    auto it = cpp_function_types.find(std::type_index(typeid(fn)));
+    if (it == cpp_function_types.end()) {
+      return PyErr_Format(PyExc_TypeError,
+          "Don't know how to create Python object for %s", typeid(fn).name());
+    }
 
-  PyTypeObject* type = (PyTypeObject*)it->second.get();
-  THPObjectPtr obj(type->tp_alloc(type, 0));
-  if (!obj) return NULL;
-  THPCppFunction* f = (THPCppFunction*)obj.get();
-  new (&f->cdata) std::shared_ptr<Function>(cdata);
-  if (!f->cdata) {
-    return NULL;
+    PyTypeObject* type = (PyTypeObject*)it->second.get();
+    THPObjectPtr obj(type->tp_alloc(type, 0));
+    if (!obj) return NULL;
+    THPCppFunction* f = (THPCppFunction*)obj.get();
+    f->cdata = cdata;
+
+    cdata->pyobj = obj.release();
+    Py_INCREF(cdata->pyobj);
```

Comment:
I think the pyobj ends up with an extra INCREF here. I think it should only have refcnt 1, but it has two (one from `tp_alloc` and one from `Py_INCREF`)

The previous code using the placement new syntax (`new (&f->cdata) std::shared_ptr<Function>(cdata);`) is better. `tp_alloc` doesn't construct the shared_ptr. It's probably fine because `tp_alloc` zeros all the memory, but I think it's better to explicitly construct it here.